### ISSUE TITLE
mintty.exe opens GUI in seperate process

### ISSUE THREAD

**Issue by P-St** [mintty/mintty#909]
Hi,

thanks for your work!

I'm already successfully embedding PuTTY's terminal window into a .NET wrapper application to provide tabs and so on. Now I tried to embed mintty, too. But it doesn't seem to work as with PuTTY.

I noticed mintty.exe always creates a new process (for the GUI?) as you can see in the output of [drstrace](http://drmemory.org/strace_for_windows.html) here:
```
PS C:\Users\pat\Desktop\DrMemory-Windows-2.2.0-1\bin> .\drstrace.exe -- C:\tools\cygwin\bin\mintty.exe
<drstrace log file is .\drstrace.mintty.exe.00292.0000.log>
<drstrace log file is .\drstrace.mintty.exe.01100.0000.log>
```
The initial starting process then exits and raises the close event in the wrapper application.

When using e.g. PuTTY, this is not the case:
```
PS C:\Users\pat\Desktop\DrMemory-Windows-2.2.0-1\bin> .\drstrace.exe -- C:\Users\pat\Desktop\putty.exe
<drstrace log file is .\drstrace.putty.exe.03552.0000.log>
```
I think this could be the reason why it is not working as with PuTTY. Is there any way to start mintty's GUI in the starting process?

Regards
Patrick

**Comment by mintty** [comment]
.net? Really? Sounds like overkill.
You may like to try out the `-d` option.

**Comment by P-St** [comment]
Thanks, I tried that already, but still there are two processes:
```
PS C:\Users\pat\Desktop\DrMemory-Windows-2.2.0-1\bin> .\drstrace.exe -- C:\tools\cygwin\bin\mintty.exe -d
<drstrace log file is .\drstrace.mintty.exe.02208.0000.log>
<drstrace log file is .\drstrace.mintty.exe.02148.0000.log>
```
Trying the ```-D``` option, which enables daemonization, there are three processes:
```
PS C:\Users\pat\Desktop\DrMemory-Windows-2.2.0-1\bin> .\drstrace.exe -- C:\tools\cygwin\bin\mintty.exe -D
<drstrace log file is .\drstrace.mintty.exe.02740.0000.log>
<drstrace log file is .\drstrace.mintty.exe.02856.0000.log>
PS C:\Users\pat\Desktop\DrMemory-Windows-2.2.0-1\bin> <drstrace log file is .\drstrace.mintty.exe.02580.0000.log>
```
I looked at the code, there is the ```child.c```. I think its function ```child_create(...)``` is used to create the terminal window. This function is called always, even with the option ```-d```

**Comment by mintty** [comment]
It is necessary to have a decoupled child processes that will then host the child program, i.e. normally the shell. I doubt that's different in putty.

**Comment by P-St** [comment]
Would it be possible to add a new option like ```--pid```, which will output the PID of the decoupled child process to the console after launching mintty? With this information I could easily embed it into the wrapper application.

**Comment by mintty** [comment]
Actually, putty apparently uses Windows CreateProcess to start the shell, not POSIX fork.
Uploaded patch with option `-Rp`.

**Comment by P-St** [comment]
Great, thanks! I tried it and it is working when running in a Cygwin Terminal:
```
$ mintty.exe -Rp
34
```
Unfortunately, when I run it from a windows console like Powershell there is no output (which is the same behaviour as when trying to get the output from the wrapper application):
```
PS C:\cygwin\bin> .\mintty.exe -Rp
PS C:\cygwin\bin>
```
Receiving log output in a windows console is not possible, too, so this seems to be a general problem:
```
PS C:\cygwin\bin> .\mintty.exe --log -
PS C:\cygwin\bin>
```
However, there is a possibility to log to a file. So, if it would be possible to write the output of ```-Rp``` to a file like e.g. ```mintty.exe -Rp --Reportfile tmp123.out```, the wrapper application could easily read the PID from this file. Or perhaps there is a more elegant way?

I hope I'm not asking too much from you!

**Comment by mintty** [comment]
Actually, that's already available, as an undocumented debug option:
`mintty -Rp --trace=FILE`

**Comment by mintty** [comment]
Or rather in this order:
`mintty --trace=FILE -R...`
For `-Rp` it does not matter but for some reporting options it does.

**Comment by P-St** [comment]
Thanks!
I tried using the ```--trace``` option and the PID was written to file. However, the actual write operation seems to be executed _after_ the Mintty window closes.
Is there a way to write the PID to file directly after the child process is created?

**Comment by BrianInglis** [comment]
Without changing `mintty`, you can find process ids yourself by looking in `/proc/`: 
```
$ grep '^/usr/bin/mintty$' /proc/*/exename
/proc/62291/exename:/usr/bin/mintty
$ grep -l '^/usr/bin/mintty$' /proc/*/exename
/proc/62291/exename
```
searching `ps` output:
```
$ ps | egrep 'mintty|PID'
      PID    PPID    PGID     WINPID   TTY         UID    STIME COMMAND
    62291       1   62291      45288  ?         197608 10:06:26 /usr/bin/mintty
```
installing Cygwin `procps-ng` package and using that `procps`: 
```
$ procps -C mintty
  PID TTY          TIME CMD
62291 ?        00:00:26 mintty
```
using the native Windows `tasklist`: 
```
$ tasklist /fi "imagename eq mintty.exe"

Image Name                     PID Session Name        Session#    Mem Usage
========================= ======== ================ =========== ============
mintty.exe                   45288 Console                   22      7,332 K
```
or the equivalent get-process lookup in **PowerShell** or **C#**. 
Just because it _could_ be done in `mintty`, does not mean that it _should_, when there are so many better alternatives than rarely used options.

**Comment by P-St** [comment]
@BrianInglis  Thanks. There is a problem with that approach. Suppose the are multiple instances of Mintty running (perhaps some instances were started not through the wrapper application):
```
$ ps | egrep 'mintty|PID'
      PID    PPID    PGID     WINPID   TTY         UID    STIME COMMAND
     1072       1    1072      12624  ?         197609 07:04:36 /usr/bin/mintty
     1063       1    1063      15608  ?         197609 07:04:29 /usr/bin/mintty
     1054       1    1054      15092  ?         197609 07:04:21 /usr/bin/mintty
     1082       1    1082      14156  ?         197609 07:05:38 /usr/bin/mintty
     1066       1    1066       9952  ?         197609 07:04:31 /usr/bin/mintty
     1060       1    1060      10064  ?         197609 07:04:27 /usr/bin/mintty
     1057       1    1057       8132  ?         197609 07:04:24 /usr/bin/mintty
     1069       1    1069      11480  ?         197609 07:04:34 /usr/bin/mintty
     1078       1    1078      11080  ?         197609 07:04:40 /usr/bin/mintty
     1051       1    1051      10164  ?         197609 07:02:57 /usr/bin/mintty
     1075       1    1075       5536  ?         197609 07:04:38 /usr/bin/mintty
```
Now the user starts a new Mintty terminal through the wrapper application. How to decide which Mintty process is the right one to embed into the wrapper application?

**Comment by mintty** [comment]
> Is there a way to write the PID to file directly after the child process is created?

Try the following patch:
```
diff --git a/src/child.c b/src/child.c
index 1215a07..3ae0bae 100644
--- a/src/child.c
+++ b/src/child.c
@@ -330,8 +330,10 @@ child_create(char *argv[], struct winsize *winp)
     exit(mexit);
   }
   else { // Parent process.
-    if (report_child_pid)
+    if (report_child_pid) {
       printf("%d\n", pid);
+      fflush(stdout);
+    }

**Comment by P-St** [comment]
Thanks, I've tried it, it's the same behaviour as without the flush command.

I've dived a little deeper into the code, and found out that the PID of the pseudo terminal is actually not what I need. When starting mintty.exe without any parameters, the PID created by ```forkpty``` will be the PID of ```bash.exe```, and not the terminal window. I'm sorry about that, should have tested this before!

I found the relevant function ```CreateWindowExW(...)``` in ```winmain.c```. I put the following code block under this function to be able to test it in the wrapper application (I don't have experience with "C"):
```
unsigned long wpid = -1;
GetWindowThreadProcessId(wnd, &wpid);
FILE *fp;  
fp = fopen("/tmp/pid.out", "w");
fprintf(fp, "%lu\n", wpid);
fclose(fp);
```
This will save the PID of the terminal window to ```/tmp/pid.out``` as soon as the window is created. I could successfully open the file in the wrapper application, read the PID and embed the window. It also works well when quickly embedding multiple instances of Mintty - e.g. when using the wrapper application's cluster mode. So I think the approach with writing the PID to file is a good solution.

It would be awesome if you could somehow integrate this into Mintty.

**Comment by mintty** [comment]
How do you actually start mintty? Wouldn't that give you access to its pid anyway?

**Comment by P-St** [comment]
I have only access to the PID of the process, which e.g. executes Mintty's main function. When this process creates another process with ```CreateWindowExW(...)```, which is the terminal window, I think there is no way to track this in the wrapper application.

I'm using the .NET System.Diagnostics.Process class to launch Mintty ([link](https://docs.microsoft.com/de-de/dotnet/api/system.diagnostics.process.start?view=netframework-4.8#System_Diagnostics_Process_Start)).

If someone knows a better way to solve this problem, I'd be glad to know it

**Comment by P-St** [comment]
Thanks, this is it! Nice work.
I can now start to integrate Mintty+Cygwin into my wrapper application. Awesome!

**Comment by mintty** [comment]
Released 3.0.3.
